fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentType.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.attachment.v1;
+
+import java.util.function.Supplier;
+
+import com.mojang.serialization.Codec;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.util.Identifier;
+
+/**
+ * An attachment allows "attaching" arbitrary data to various game objects (entities, block entities, worlds and chunks at the moment).
+ * Use the methods provided in {@link AttachmentRegistry} to create and register attachments.
+ *
+ * <p>Attachments can optionally be made to persist between restarts using a provided {@link Codec}.</p>
+ *
+ * @param <A> type of the attached data. It is encouraged for this to be an immutable type.
+ */
+@ApiStatus.NonExtendable
+@ApiStatus.Experimental
+public interface AttachmentType<A> {
```

Comment:
I think that more attention should be given to entity copies.
1. For NeoForge, I added a `copyOnDeath` boolean to optionally copy a serializable attachment on player or entity death.
2. We should automatically copy serializable attachments when a player is cloned due to non-death circumstances, e.g. when returning from the end.

Reply:
Why should this be automatic? I understand that this may seem convenient, but vanilla doesn'tcopy all the data and we have `AFTER_RESPAWN`/`COPY_FROM`/`AFTER_ENTITY/PLAYER_CHANGE_WORLD`

Reply:
And yes, attachments don't only apply to entities